### ISSUE TITLE
(umbrella issue) Define, Deprecate, Roadmap the NodeNetworkUnvailable / NetworkUnavailable:false taints, tolerations, conditions - including CNI/CloudController reccomendations

### ISSUE THREAD

**Issue by jayunit100** [kubernetes/kubernetes#120486]
TLDR,  There are many moving parts in a K8s cluster that may use / depend on 
- the  `"node.kubernetes.io/network-unavailable" condition`  and/or
- `v1.TaintNodeNetworkUnavailable taint` Condition/Taint couple 

In ambiguous/ad-hoc sorts of ways.  This makes it confusing for users and vendors to know when they should/shouldnt manipulate the values of such TCs.  

## Solution

- We may need to make more well defined taints/tolerations that say things precisely about network state.  This might break systems relying heavily on the oversimplified "NetworkUnavailable" semantics we currently have.
- We need to deprecate, re-define, and document taints/tolerations for k8s networks...  This is a several week or month process b/c it will require lots of water carrying across vendors / sig-docs/sig-net/sig-node.

*By the way, lots of water-carrying here between vendors/sig-net, might be a good "first" issue for someone whose very eager to learn about the history/semantics of K8s networking and pod scheduling, and has many weeks to dedicate !*

## Detailed Summary

(9/12 Editing this issue to update it per comments...)

Many, many issues expose the complexity of node networking readiness.
- https://github.com/antrea-io/antrea/issues/2298
- https://github.com/flannel-io/flannel/issues/1431
- https://github.com/projectcalico/calico/issues/6165
- https://github.com/projectcalico/calico/issues/5233
- https://github.com/containernetworking/cni/issues/859
- https://github.com/containernetworking/cni/issues/628

There's proposals to get **CNI Status** into the CNI specification.  For now, the most urgent need appears to be to fully 
- DEFINE the **current taints, tolerations, and expectations** that CNI, cloud provider, and others should have
- FILE **bugs** against Kuberentes and CNI/cloud provider implementations that may/may not use the taints/tolerations in Kubernetes properly
- TALK to the Multi-networking and CNI providers individually to figure out what the lowest common denominator for node networking readiness and tainting is
- WRITE a **kep** to update all of these taints, tolerations, and so on so that they are consistently aligned to the indended use (and deprecate the taints which don't have a usable definition once we consider this broader world view of K8s networking)
- Original brainstorming doc... not relevant anymore but for historical reasons:  https://docs.google.com/document/d/1xlZYVw2iu447b5Us8V8gRC5uigjrjtOCX7RgQ19Oe4/edit

# Goals

I think the goals of this issue are something like: 

- The CNI and CRI/Kubelet would have a *well defined "taint namespace"* (not literally, but, some sort of semantic boundary)... related specifically to when CNIs were ready to make pods
- There would be **comprehensive guidance for CNIs and cloud controllers** on how / when to/ when not to modify these taints conditions
- Taints/Conditions related to network readiness, would have clear well-defined meaning in **multi-networking and multi-OS** environments w/ docs on them
- There would be no **cloud specific behaviour** in K8s related to these taints and conditions
- **Maybe this means we need to entirely rethink the value of high-level, networking related taints**

```[tasklist]
### Tasks
- [ ] https://github.com/kubernetes/website/issues/43018
- [ ] https://github.com/kubernetes/website/issues/42917
```

**Comment by jayunit100** [comment]
https://github.com/antrea-io/antrea/issues/2298
https://github.com/flannel-io/flannel/issues/1431
https://github.com/projectcalico/calico/issues/6165 
https://github.com/projectcalico/calico/issues/6165 
https://github.com/projectcalico/calico/issues/5233

**Comment by jayunit100** [comment]
https://github.com/kubernetes/website/issues/42917

**Comment by jayunit100** [comment]
cc @dims who i think is currentyl doing cloudprovider out of tree stuff, i wonder if that will effect the conditions at all and how theyre updated and make this more / less confusing in the future?

**Comment by danwinship** [comment]
AFAIK, the situation is this:

- (Just to be clear: the taint is called `"node.kubernetes.io/network-unavailable"` or `v1.TaintNodeNetworkUnavailable`, and the condition is called `"NetworkUnavailable"` or `v1.NodeNetworkUnavailable`. That is, both of them have "`Node`" in the name of their constant in `"k8s.io/api/core/v1"`, but not in the _value_ of the constant.)
- The taint simply mirrors the condition, as per https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-nodes-by-condition
- kubelet will set the `NodeNetworkUnavailable` condition to `True` at startup if you are on GCE (and _only_ if you are on GCE). AFAIK nothing else in core k8s else ever sets the condition to `True`.
- The NodeIPAM controller in kcm will set the `NodeNetworkUnavailable` condition to `False` when it allocates a CIDR for that node, whether or not it was previously set `True`. AFAIK nothing else in core k8s ever sets the condition to `False`.

So:
- If you are on GCE, and you are using the NodeIPAM controller, then Nodes will have the NodeNetworkUnavailable condition/taint when they are created, and it will be cleared once the Node is able to assign IPs to pods.
- If you are _not_ on GCE, and you are using the NodeIPAM controller, then Nodes will never have a `True` NodeNetworkUnavailable condition, or the NodeNetworkUnavailable taint, but they will eventually get a `False` NodeNetworkUnavailable condition once they have been assigned a PodCIDR.
- If you are _not_ on GCE, and _not_ using the NodeIPAM controller, then no core k8s code will ever create or modify the NodeNetworkUnavailable condition or taint.
- If you are on GCE and _not_ using the NodeIPAM controller, then nodes will be tainted when they are created, and nothing in core k8s will ever un-taint them. This is why various CNI plugins have code to unset the condition themselves; because if they didn't, then they wouldn't work on GCE, because Nodes would remain tainted forever.

This is clearly wrong but no one has ever bothered to try to fix it.

Note that nodes will also have a `Ready` condition of `False` (and thus also a `"node.kubernetes.io/not-ready"` taint) until CRI is happy with the CNI situation, so in "heavy network plugin" setups, it is generally the NodeNotReady taint that blocks pod-network pods from starting too early, not the NodeNetworkUnavailable taint.

**Comment by sftim** [comment]
We have a node configuration field `registerWithTaints`; it'd (IMO) be useful to have a similar field `startupWithTaints`.

I would also default that field to `["node.kubernetes.io/not-ready","node.kubernetes.io/network-unavailable"]` _and_ have the kubelet infer that this means it should start with the relevant conditions applied.

We could go further, eg:
- log something if you override the list of startup taints so that `node.kubernetes.io/not-ready` is not included
- `startupWithExtraTaints`, an optional field to specify additional taints only.

**Comment by jayunit100** [comment]
gotcha, ok, in my case, - podCidr exists but no condition....  not on GCE.
```
kubo@uOFLhGS9YBJ3y:~/geetika$ kubectl get node windows-cluster-md-0-8fvd6-7dd7b779f4x5xk8x-9k82j -o yaml | grep -i podCIDR                                                                                                                                                                 podCIDR: 100.96.87.0/24                                                                                                                                                                                                                                                                
  podCIDRs:                                                                                                                                                                                                                                                                              kubo@uOFLhGS9YBJ3y:~/geetika$ kubectl get node windows-cluster-md-0-8fvd6-7dd7b779f4x5xk8x-9k82j -o yaml | grep -i Network                                                                                                                                                               
```

**Comment by danwinship** [comment]
> We have a node configuration field `registerWithTaints`; it'd (IMO) be useful to have a similar field `startupWithTaints`.
> 
> I would also default that field to `["node.kubernetes.io/not-ready","node.kubernetes.io/network-unavailable"]` _and_ have the kubelet infer that this means it should start with the relevant conditions applied.

So that sounds more like `startupWithConditions` than `startupWithTaints`...

The problem with trying to reuse the existing NetworkUnavailable condition/taint here is that existing components (kcm, network plugins) already have logic to remove that condition, in circumstances that don't always correspond to the network actually being ready. I think we'd want to deprecate the old condition and create a new one with a new name and clearer semantics.

Of course, you'd also have to ensure that kubelet was configured with that `startupWithConditions: ...`  if you had a CNI plugin that supported removing the condition, and _not_ configured with that option if you didn't. Except, also, ideally you don't want the CNI plugin to remove the condition itself anyway, because [the ability to modify Node objects is useful to attackers](https://www.paloaltonetworks.com/resources/whitepapers/kubernetes-privilege-escalation-excessive-permissions-in-popular-platforms) and there's a special kubelet-specific hack that makes it less dangerous if we let kubelet modify the Node rather than giving another component permission to do it.

So really, what we want is that kubelet starts up, without needing any extra new config, and CRI/CNI somehow tells it to apply the condition/taint, and then later CRI/CNI tells it to remove the condition/taint.

It would also be nice if we could, somehow, have a taint that applied to pod-network pods, but _not_ host-network pods, which is what we really want here, though I don't think we can really do that without adding special hacks in kube-scheduler...

**Comment by jayunit100** [comment]
I feel like theres too much drift around CNI, cloud provider, and OS to really force any one "heres how we should taint" scenario, but I don't know if i have enough data to back this up.

**Comment by jayunit100** [comment]
I think with appropriate guidance a newcomer or not too-technical person good gather this info from stake holders by going to various sig-meetings and polling people.  It would require several months of commitment, but I dont think from a "code" perspective its a super advanced problem.  I'd like to propose this as a good CNCF onboarding task for someone new to get involved and make an impact making K8s easier to understand and simpler to reason about.

 cc @knabben  @shaneutt

**Comment by jayunit100** [comment]
hi folks, Ive tried to redefine this issue so its more actionable but its tricky.  Let me know if anyone thinks i should keep it, or re-file it, or something else.

**Comment by liljenstolpe** [comment]
This gets to be more interesting with multi-networking.  Which network(s) are available/unavailable.  Example could be that the primary network is up, but one of the secondaries is not available, but the particular secondary is required for a given pod (say a 5G RAN service pod).

**Comment by danwinship** [comment]
We definitely don't want to taint the whole node because a _secondary_ network isn't available. But I guess having a way to indicate "this node can't currently accept pods that want to use Network X" would work (where "Network X" might be the default network or a secondary).

Especially in that case, taints may not be the right model.

**Comment by jayunit100** [comment]
> taints may not be the right model... 
 
If we could go back in time, wouldn't we  rather just have lots and lots of labels, and have controllers or something decide when a combination of labels was sufficiently red to taint a node? 

```
version: 1.x
Name: LabelToTaintConfig
spec:
   must-be-true:
   - calico.networking-cni*
   - k8s.io.*
```
and then
```
for N in nodes:
  for L in node.labels:
     if L in must-be-true:
        # ... this example was python until just now... 
        log("warning! im tainting node %v bc %v", n, x)
        # ... ok pretending python again....
        taint_the_node(n, x)
     else:
        log("i found a label %v but im not tainting this node b/c its not in the taint list",L)
```

**Comment by sftim** [comment]
Watch out though that of course we can't go back in time. We do have existing end user and integration expectations around the existing condition and existing taint.

Having _two_ conditions not one may well not help with end user understanding as much as we'd hoped; we'd need a transition plan that ensures people end up less confused, not more.

**Comment by jayunit100** [comment]
> We do have existing end user and integration expectations around the existing condition and existing taint.

It sounds like theres two "requirements" here:

- @danwinship's above " I think we'd want to deprecate the old condition and create a new one with a new name and clearer semantics."
- Have the kubelet start up, without needing any extra new config, and CRI/CNI somehow tells it to apply the condition/taint, and then later CRI/CNI tells it to remove the condition/taint.
- Have a taint that applied to pod-network pods, which the scheduler should be capable of interpreting **WHICH DOES NOT** prevent host-network pods from running
- @sftim's "we'd need a transition plan that ensures people end up less confused, not more."

Are there any other requirements folks want to add ?

**Comment by MikeZappa87** [comment]
@jayunit100 this is something that the KNI is looking at. In summary we aim to provide status, availability and metrics for network(s) since the current approach is rather simplistic.

**Comment by danwinship** [comment]
So this is the sort of thing we ought to be fixing as part of [KNI](https://github.com/kubernetes/enhancements/issues/4410)
